### PR TITLE
Integrate new machinecontroller types

### LINKED ISSUE #2031: Integrate new cluster machine types into Kubermatic
**User Story**

As a user of Kubermatic I want it to use the same types as the current machine-controller in order to be able to update the machine-controller.

**Acceptance criteria**

* Kubermatic uses the same machine types as the current (v0.9.0) version of the machine-controller

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2031 

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
Kubermatic now uses the types from the [cluster api project](https://github.com/kubernetes-sigs/cluster-api) to manage nodes
```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
wow because of the change !
wow because this pull is massive !

in general it changes the components to use `clusterv1alpha1.Machine` resource ?

File: api/pkg/resources/machine/machine.go
Comment (alvaroaleman) [inline api/pkg/resources/machine/machine.go]: @p0lyn0mial Do you know what did that defaulting before and why it doesnt happen anymore?

**Comment by alvaroaleman** [comment]
> in general it changes the components to use clusterv1alpha1.Machine resource ?

@p0lyn0mial Yes exactly

File: api/pkg/resources/machine/machine.go
Comment (p0lyn0mial) [reply api/pkg/resources/machine/machine.go]: before `ContainerRuntime.Name` was hardcoded in the test's payload, see: https://github.com/kubermatic/kubermatic/blob/master/api/pkg/handler/node_v1_test.go#L578

after your changes the payload has been removed, https://github.com/kubermatic/kubermatic/pull/2056/files/942007868a2c9fc6712e7555d06c799c2eda5ff6#diff-0534c634e6bf2edb16be53d744b5aa00L578 

I'm not sure if there was `defaulting` logic before.

**Comment by alvaroaleman** [comment]
@p0lyn0mial So that test was just me being stupid, I did not intend to change the test itself. Thanks for pointing out the mistake!

File: api/pkg/controller/ipam/controller.go
Comment (p0lyn0mial) [inline api/pkg/controller/ipam/controller.go:275]: you cannot find the machine when using the lister ?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Yes, this broke when swapping the lister to use the one from cluster api. I have no idea why. In the machine-controller the very same listers are used and work just fine. Got any idea what the issue could be?

This currently blocks the PR :/

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: mhm, that's really strange, a few lines earlier we actually retrieve the object from the lister
```
namespace, name, err := cache.SplitMetaNamespaceKey(key)
if err != nil {
	return err
}
listerMachine, err := c.machineLister.Machines(namespace).Get(name)
```

then we pass it to `syncMachine` which passes it to `initMachineIfNeeded` which eventually passes it to `awaitIPSync` method.

Is the object on the list returned from `machineLister.List()` ?
Could you tell me the namespace and the name of the object we are trying to get ?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Yes, the list works fine and contains all expected machines.
To me it looks like the `Update` has no effect on the machine in the lister:

https://github.com/kubermatic/kubermatic/pull/2056/files#diff-2163c4c117f48eddce87cf9c852d8caaR172

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: oh, so you actually can get the object from the lister but it contains stale data ?
this line `return cfg2.Network != nil && cfg2.Network.CIDR == cidr, nil` ?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Yes, exactly

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Got it, allthought it doesnt make me happy:

Cluster-api uses a vendored version of client-gen to generate its client: https://github.com/kubernetes-sigs/cluster-api/blob/7db708b5df567c79e2e46f9c37cdbf001bef1302/Makefile#L40

That version seems not to properly work with the listers. If I use my locally installed copy of `client-gen` by commenting out the build step in the cluster apis Makefile, then copy the generated fake client over, the tests pass

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: oh man, that's suspicious, aren't they using `release-1.9` ?
what's your local version ?

https://github.com/kubernetes-sigs/cluster-api/blob/7db708b5df567c79e2e46f9c37cdbf001bef1302/Gopkg.toml#L82

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Yup, its this patch that is missing: https://github.com/kubernetes/kubernetes/pull/57504/files#diff-725091da93c11d914fc1c0a3eb064a9a

Got added in 1.10. They wont update the client thought, as they depend on the deprecated apiserver-builder which uses 1.9. Also, just updating the code-gen within the project is not possible, since the above linked pull also changes internal interface within client-go: https://github.com/kubernetes/kubernetes/pull/57504/files#diff-1fedd5b881e8fbda5aaefa9bd74b359bR65

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: @mrIncompetent How do you feel about generating our own clientset?

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: @alvaroaleman the linked patch changes the fake client not the real one :) I have been thinking that the real lister is broken :)

If this only concerns the test perhaps we could add a nasty dirty flag just for the tests that would switch between lister and the normal client :) 🥇 ?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: >  [...] flag just for the tests that would switch between lister normal client :) ￼ ?

What do you mean by `switch between lister normal client`?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: Ah you mean within the controller...Thats very, very ugly.

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: yeah, I know :) That would be temporal until upstream catches up.

Comment (p0lyn0mial) [reply api/pkg/controller/ipam/controller.go:275]: @alvaroaleman or maybe we could fix the tests ? I think that they are running informers which doesn't sound good, they could only use the indexer ?

Comment (mrIncompetent) [reply api/pkg/controller/ipam/controller.go:275]: > @mrIncompetent How do you feel about generating our own clientset?

Whats the strategy you have in mind here? And whats the effort?

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: @p0lyn0mial Unfortunatelly that wont work because the controller updates stuff via the client and that wont get reflected in the indexer, so we are back to square 0

@mrIncompetent The idea was to write a simple bashscript that generates a clientset for the cluster-api from the upstream types by using a newer version of `client-gen`, copy out just the `fake` part of it, put it somewhere within Kubermatic and use that for the tests

Comment (alvaroaleman) [reply api/pkg/controller/ipam/controller.go:275]: `make test` will now generate a clientset that will get cleaned up after the execution of the ipam controller tests: 126515327d9bb1af6b9c0d85e848ed53e1521e28

File: api/Gopkg.lock
Comment (mrIncompetent) [inline api/Gopkg.lock:173]: I guess thats caused by upstream cluster-api ?

File: api/pkg/controller/cluster/pending.go
Comment (mrIncompetent) [inline api/pkg/controller/cluster/pending.go]: Definictions -> Definitions

File: api/pkg/controller/cluster/userclusterresources.go
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go:407]: use `resources.DeepEqual` here

File: api/pkg/controller/cluster/userclusterresources.go
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]: Can we use consistently use `CustomResourceDefinition` instead of `custom resource definition` across errors and logs?

File: api/pkg/controller/cluster/userclusterresources.go
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]: https://github.com/kubermatic/kubermatic/pull/2056/files#r220087763

File: api/pkg/controller/cluster/userclusterresources.go
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]: https://github.com/kubermatic/kubermatic/pull/2056/files#r220087763

File: api/pkg/controller/ipam/controller.go
Comment (mrIncompetent) [reply api/pkg/controller/ipam/controller.go:275]: sgtm

File: api/Gopkg.lock
Comment (alvaroaleman) [reply api/Gopkg.lock:173]: Yes

**Comment by alvaroaleman** [comment]
@mrIncompetent Comments addressed, PTAL

**Review by mrIncompetent — DISMISSED** [review]

**Comment by alvaroaleman** [comment]
I'll keep this open until kubermatic/machine-controller#326 is merged, as that updates the cluster-api dependency to not have that change in kubermatic git history

**Comment by alvaroaleman** [comment]
@p0lyn0mial Can you take a look on this PR and check if you are fine with it?

**Comment by alvaroaleman** [comment]
@p0lyn0mial @mrIncompetent kubermatic/machine-controller#326  got merged and this PR updated. From my POV this is done, PTAL

**Review by mrIncompetent — DISMISSED** [review]

**Review by kron4eg — APPROVED** [review]

**Comment by p0lyn0mial** [comment]
@alvaroaleman nice pull, LGTM

**Review by p0lyn0mial — APPROVED** [review]

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**
